### PR TITLE
Time Entries to Typescript

### PR DESCRIPTION
**Pull request by larstroost**

### CODE REVIEW

File: ducks/time-entries.ts [hunk 1]
```diff
@@ -40,14 +60,33 @@ const timeEntriesActiveSortDirectionSelector = createSelector(
...
+      (client) => item.clientId === client.id
+    );
+    const clientName = clientMatch && clientMatch.clientName;
+    return { ...item, clientName };
+  })
+);
+
+export const timeEntriesFilterClientsSelector = createSelector(
+  clientsItemsSelector,
+  (items) => items.reduce((clientNames, item) => ([...clientNames, {
```
Comment (boris-arkenaar) [inline ducks/time-entries.ts]: Use `map` instead of `reduce`.

File: ducks/time-entries.ts [hunk 2]
```diff
@@ -40,14 +60,33 @@ const timeEntriesActiveSortDirectionSelector = createSelector(
...
 export const timeEntriesSelector = createSelector(
-  [timeEntriesItemsSelector,
+  [timeEntriesWithClientNamesSelector,
     timeEntriesActiveFilterSelector,
     timeEntriesActiveSortBySelector,
     timeEntriesActiveSortDirectionSelector
   ],
   (items, activeFilter, activeSortBy, activeSortDirection) => (
-    (items.filter((item) => !activeFilter || item.client === activeFilter)
+    (items.filter((item) => !activeFilter || item.clientId === activeFilter)
```
Comment (boris-arkenaar) [inline ducks/time-entries.ts]: You can remove the braces around `items.filter`.

File: components/time-entries/TimeEntries.tsx [hunk 3]
```diff
@@ -1,16 +1,39 @@
...
 
-class TimeEntries extends React.Component {
+interface TimeEntry {
+  clientName?: string;
+  startTime: string;
+  stopTime: string;
+  id: string;
+}
+
+export interface Clients {
```
[the remark this answers is not in the document]
Comment (larstroost) [reply components/time-entries/TimeEntries.tsx]: Even if the duck of clients is a .js file? Of do I then need to define it in the time-entries.ts?

[the remark this answers is not in the document]
Comment (boris-arkenaar) [reply components/time-entries/TimeEntries.tsx]: Make that file a `.ts` file as well then.

File: components/time-entry/TimeEntry.tsx [hunk 4]
```diff
@@ -1,51 +1,56 @@
...
+import { TimeEntryModel } from '../../ducks/time-entries';
 
 const isoToLocale = (time) => new Date(time).toLocaleTimeString('nl-NL', { hour: '2-digit', minute: '2-digit' });
 
 interface TimeEntryProps {
-    clientName?: string;
-    startTime: string;
-    stopTime: string;
-    id: string;
+    clientName?: TimeEntryModel["clientName"];
```
Comment (boris-arkenaar) [inline components/time-entry/TimeEntry.tsx]: Try to define it like

    timeEntry: TimeEntryModel;

File: components/time-entry/TimeEntry.tsx [hunk 5]
```diff
@@ -1,51 +1,56 @@
...
-            - Date.parse(startTime))
-            - 1 * 1000 * 60 * 60)}
-        </p>
-      </div>
-    </div>
-  </section>
-);
-
-(TimeEntry as any).defaultProps = {
+class TimeEntry extends React.Component<TimeEntryProps, null> {
```
Comment (boris-arkenaar) [inline components/time-entry/TimeEntry.tsx]: You could leave the second type definition

    React.Component<TimeEntryProps>

**Pull request merged**